onchain/rollups/contracts/consensus/quorum/Quorum.sol

```diff
@@ -0,0 +1,133 @@
+
+    /// @notice For each claim, the set of validators that agree
+    /// that it should be submitted to the history contract.
+    mapping(bytes => EnumerableSet.AddressSet) internal yeas;
+
+    /// @notice An ERC-20 token transfer failed (`transfer` returned `false`).
+    error ERC20TokenTransferFailed();
+
+    /// @notice Construct a Quorum consensus
+    /// @param _validators the list of validators
+    constructor(IHistory _history, address[] memory _validators) {
+        // Iterate through the array of validators,
+        // and grant to each the validator role.
+        for (uint256 i; i < _validators.length; ++i) {
+            grantRole(VALIDATOR_ROLE, _validators[i]);
+        }
+
+        // Set history.
+        history = _history;
+    }
+
+    /// @notice Get the history contract.
+    /// @return The history contract
+    function getHistory() external view returns (IHistory) {
+        return history;
+    }
+
+    /// @notice Get a claim from the current history.
+    ///         The encoding of `_proofContext` might vary depending on the
+    ///         implementation of the current history contract.
+    /// @inheritdoc IConsensus
+    function getClaim(
+        address _dapp,
+        bytes calldata _proofContext
+    ) external view override returns (bytes32, uint256, uint256) {
+        return history.getClaim(_dapp, _proofContext);
+    }
+
+    /// @notice Submits a claim for voting.
+    ///         If this is the claim that reaches the majority, then
+    ///         the claim is submitted to the history contract.
+    ///         The encoding of `_claimData` might vary depending on the
+    ///         implementation of the current history contract.
+    /// @param _claimData Data for submitting a claim
+    /// @dev Can only be called by a validator,
+    ///      and the `Quorum` contract must have ownership over
+    ///      its current history contract.
+    function submitClaim(
+        bytes calldata _claimData
+    ) external onlyRole(VALIDATOR_ROLE) {
+        // Get the set of validators in favour of the claim
+        EnumerableSet.AddressSet storage claimYeas = yeas[_claimData];
+
+        // Add the message sender to such set.
+        claimYeas.add(msg.sender);
+
+        // Get number of validators in favour of the claim.
+        uint256 numOfVotesInFavour = claimYeas.length();
+
+        // Get the number of validators in the quorum.
+        uint256 quorumSize = getRoleMemberCount(VALIDATOR_ROLE);
+
+        // If this claim already has half of the quorum's approval,
+        // then we can submit it to the history contract.
+        if (numOfVotesInFavour > quorumSize / 2) {
+            history.submitClaim(_claimData);
+        }
+    }
+
+    /// @notice Equally share all tokens from some ERC-20 contract
+    /// amongst all validators in the quorum.
+    /// @param _token The token contract
+    function shareERC20Tokens(IERC20 _token) external {
```

Comment:
I know that we have a limited set of validators, but I'm still not comfortable pushing transfers like this in a for loop.

Maybe we can use the [payment splitter](https://docs.openzeppelin.com/contracts/4.x/api/finance#PaymentSplitter) from Open Zeppelin.

Reply:
True, I will take a look into that. Thanks!